src/index.ts

```diff
@@ -268,3 +269,15 @@ async function closeGracefully (signal: number): Promise<void> {
   // eslint-disable-next-line @typescript-eslint/no-misused-promises
   process.once(signal, closeGracefully)
 })
+
+const uncaughtHandler = (error: any): void => {
+  log.error('Uncaught Exception:', error)
+  if (ALLOW_UNHANDLED_ERROR_RECOVERY && (RECOVERABLE_ERRORS === 'all' || RECOVERABLE_ERRORS.includes(error?.code) || RECOVERABLE_ERRORS.includes(error?.name))) {
+    log.trace('Ignoring error')
```

Comment:
This isn't considered best practice - https://nodejs.org/api/process.html#warning-using-uncaughtexception-correctly

Reply:
I understand, but we need some kind of error recovery instead of allowing things to just die like they were in Tiros. We could probably default this to FALSE and add a warning in the readme. Tiros needs this and probably some restarting of the server (to follow best practices). 


> Unhandled exceptions inherently mean that an application is in an undefined state. 

I started to just listen for `ERR_STREAM_PREMATURE_CLOSE` which we *know* is a recoverable state. "in an undefined state" is not true in this instance.  

this change allows us to recover from *anything* which could cause problems in the future:

There is a case where some error in libp2p/helia/fastify/helia-server.ts could happen that is unrecoverable, and this does an infinite loop of "On error resume next" and we don't find out until money has been eaten up running a dead service.


However, we still need to block this server from dying in instances where we know we can safely recover, and unblocking Tiros was foremost on my mind.

Reply:
The point of the linked warning is that you can't tell if you can safely recover so the only safe thing you can do is exit the process and restart.

If we have unhandled exceptions being thrown, these are bugs that should be fixed.

Reply:
I agree they're bugs that should be fixed, but should we allow a server to die given underlying libraries bugs if they're recoverable? Given the expectation of helia-http-server, I don't think so. It's supposed to fetch content using helia/libp2p, and return a result. If they die when fetching, we should certainly return a 500 error instead, and recover, right?

BTW, The only other place I saw `ERR_STREAM_PREMATURE_CLOSE` was in https://github.com/ChainSafe/js-libp2p-yamux, which is listening for those errors, so i don't think the uncaught exception is coming from there. it's likely coming from somewhere else in the libp2p stack.

Reply:
I just realized we should add a listener on the req and resp stream for ERR_STREAM_PREMATURE_CLOSE...

Reply:
updated https://github.com/ipfs/helia-http-gateway/issues/112

Reply:
> should we allow a server to die given underlying libraries bugs if they're recoverable?

Again, the point of the linked warning is that you can't tell if you can safely recover.

Consider something like this:

```js
process.on('uncaughtException', (err) => {
  if (err.message === 'recoverable') {
    // it's ok, it's recoverable
    return
  }

  console.error(err)
  process.exit(1)
})

const fd = await fs.open('/path/to/file.txt')

// something that causes an error
throw new Error('recoverable')

fd.close()
```

It looks recoverable, but the file descriptor is never closed so it leaks memory.

Basically if you're in an uncaught exception handler all bets are off.